oauth2_http/java/com/google/auth/oauth2/OAuth2Credentials.java

```diff
@@ -144,75 +152,182 @@ public void getRequestMetadata(
    */
   @Override
   public Map<String, List<String>> getRequestMetadata(URI uri) throws IOException {
-    synchronized (lock) {
-      if (shouldRefresh()) {
-        refresh();
-      }
-      if (requestMetadata == null) {
-        throw new NullPointerException("requestMetadata");
-      }
-      return requestMetadata;
-    }
+    return unwrapDirectFuture(asyncFetch(MoreExecutors.directExecutor())).requestMetadata;
   }
 
-  /** Refresh the token by discarding the cached token and metadata and requesting the new ones. */
+  /**
+   * Request a new token regardless of the current token state. If the current token is not expired,
+   * it will still be returned during the refresh.
+   */
   @Override
   public void refresh() throws IOException {
+    AsyncRefreshResult refreshResult = getOrCreateRefreshTask();
+    refreshResult.executeIfNew(MoreExecutors.directExecutor());
+    unwrapDirectFuture(refreshResult.task);
+  }
+
+  /**
+   * Refresh these credentials only if they have expired or are expiring imminently.
+   *
+   * @throws IOException during token refresh.
+   */
+  public void refreshIfExpired() throws IOException {
+    // asyncFetch will ensure that the token is refreshed
+    unwrapDirectFuture(asyncFetch(MoreExecutors.directExecutor()));
+  }
+
+  /**
+   * Attempts to get a fresh token.
+   *
+   * <p>If a fresh token is already available, it will be immediately returned. Otherwise a refresh
+   * will be scheduled using the passed in executor. While a token is being freshed, a stale value
+   * will be returned.
+   */
+  private ListenableFuture<OAuthValue> asyncFetch(Executor executor) {
+    AsyncRefreshResult refreshResult = null;
+
+    // Schedule a refresh as necessary
     synchronized (lock) {
```

Comment:
I wonder Is it still necessary to go inside the lock if the token state `Fresh`? Is it viable to just do the following:
```
if (getState() == CacheState.Fresh) {
    return Futures.immediateFuture(value);
}
```

Reply:
Thats good observation, I added the optimization. Thanks!